src/Drivers/Support/SharedDriverLogicHandler.php

```diff
@@ -0,0 +1,359 @@
+            }
+        }
+
+        ksort($conversions);
+
+        return $conversions;
+    }
+
+    /**
+     * Return the converted values between the $from and $to parameters. If no $date
+     * parameter is passed, we use today's date instead.
+     *
+     * @param  int  $value
+     * @param  string  $from
+     * @param  string|string[]  $to
+     * @param  Carbon|null  $date
+     * @return float|array<string, float>
+     *
+     * @throws InvalidDateException
+     * @throws InvalidCurrencyException
+     * @throws RequestException
+     * @throws InvalidArgumentException
+     */
+    public function convert(int $value, string $from, string|array $to, Carbon $date = null): float|array
+    {
+        if (is_string($to)) {
+            return (float) $this->exchangeRate($from, $to, $date) * $value;
+        }
+
+        $exchangeRates = $this->exchangeRate($from, $to, $date);
+
+        foreach ($exchangeRates as $currency => $exchangeRate) {
+            $exchangeRates[$currency] = (float) $exchangeRate * $value;
+        }
+
+        return $exchangeRates;
+    }
+
+    /**
+     * Return an array of the converted values between the given date range.
+     *
+     * @param  int  $value
+     * @param  string  $from
+     * @param  string|string[]  $to
+     * @param  Carbon  $date
+     * @param  Carbon  $endDate
+     * @return array<string, float>|array<string, array<string, float>>
+     *
+     * @throws InvalidCurrencyException
+     * @throws InvalidDateException
+     * @throws RequestException
+     * @throws InvalidArgumentException
+     */
+    public function convertBetweenDateRange(
+        int $value,
+        string $from,
+        string|array $to,
+        Carbon $date,
+        Carbon $endDate,
+    ): array {
+        $exchangeRates = $this->exchangeRateBetweenDateRange($from, $to, $date, $endDate);
+
+        $conversions = [];
+
+        if (is_array($to)) {
+            foreach ($exchangeRates as $date => $exchangeRate) {
+                foreach ($exchangeRate as $currency => $rate) {
+                    $conversions[$date][$currency] = (float) $rate * $value;
+                }
+            }
+
+            return $conversions;
+        }
+
+        foreach ($exchangeRates as $date => $exchangeRate) {
+            $conversions[$date] = (float) $exchangeRate * $value;
+        }
+
+        return $conversions;
+    }
+
+    /**
+     * If the 'from' and 'to' currencies are the same, we don't need to make a request to
+     * the API. Instead, we can build the response ourselves to improve the performance.
+     *
+     * @param  Carbon  $startDate
+     * @param  Carbon  $endDate
+     * @return array<string, float>
+     */
+    private function exchangeRateDateRangeResultWithSameCurrency(
+        Carbon $startDate,
+        Carbon $endDate,
+    ): array {
+        $conversions = [];
+
+        for ($date = clone $startDate; $date->lte($endDate); $date->addDay()) {
```

Comment:
In the scenario that someone passes in the same _date_ but with different _times_, is this going to cause 2 results to be returned? (I initially thought different microseconds but any time change seems like it'd cause this result).

Reply:
That's actually a great question! I hadn't ever considered supporting multiple times during the same day. When I first built the package, I'd only imagined it being used for exchange rates. I think I'll need to take a look at this bit and make sure the package properly supports multiple exchange rates for the same day 🙂

Reply:
Ah, I wasn't clear in my first comment - apologies! I meant more that if a user did:

```php
$dateStart = // some time retrieved from somewhere as 2022-01-01 12:00:00
$dateEnd = // some time retrieved from somewhere as 2022-01-01 12:00:01
->exchangeRateBetweenDateRange('GBP', 'USD', $dateStart, $dateEnd)
```

In that scenario, the start is less than the end, but that'll cause a second day to be provided.

I guess the question is: should the package be concerned about the correctness of those dates, or push that responsibility back to the developer?

Reply:
Ahh sorry, yeah I misunderstood haha. I reckon I'll just push that responsibility back to the developer. I don't think anyone's complained about this bit so far, but if it is causing any issues, I'd be happy to maybe consider handling it on our end instead 🙂